plugins/module_utils/ansible_freeipa_module.py

```diff
@@ -579,6 +580,59 @@ def names(self):
         def __getattr__(self, name):
             return self.get(name)
 
+        def get_ipa_command_args(self, **kwargs):
+            """
+            Return a dict to be passed to an IPA command.
+
+            The keys of ``ipa_param_mapping`` are also the keys of the return
+            dict.
+
+            The values of ``ipa_param_mapping`` needs to be either:
+                * A str with the name of a defined method; or
+                * A key of ``AnsibleModule.param``.
+
+            In case of a method the return of the method will be set as value
+            for the return dict.
+
+            In case of a AnsibleModule.param the value of the param will be
+            set in the return dict. In addition to that boolean values will be
+            automaticaly converted to uppercase strings (as required by FreeIPA
+            server).
+
+            """
+            args = {}
+            for ipa_param_name, param_name in self.param_mapping.items():
+
+                # Check if param_name is actually a param
+                if param_name in self.ansible_module.params:
+                    value = self.ansible_module.params_get(param_name)
+                    if isinstance(value, bool):
+                        value = "TRUE" if value else "FALSE"
+
+                # Since param wasn't a param check if it's a method name
```

Comment:
I think this param_name as method needs to be reworked. In ipadnszone for example there are a lot of checks in the methods for param_names. The general validation checks should be executed before the connection to IPA has been established.
Do we have cases where a getter is needed later on with an IPA connection? Can we get rid of param_names as methods?

Reply:
I agree that the validation should be done before connecting to IPA API, but sometimes the validation have to be done after that, and after an object is retrieved, for example, in `ipavault` (see `check_encryption_params`). It is a corner case, but we have a few corners.

The checks in `ipadnszone` should be fixed in `ipadnszone`. I'd have to review them, but had no time to do it, yet.

As of not having `methods` (or what I prefer `callables`) as `param_names`, as we discussed offline, I believe the main issue is easily differentiating from playbook parameters (`strings`) and callables. I have changed the code to use the callable itself, instead of a string of its name.

Let me know what you think about this change.

Reply:
Yes, that is more obvious than the old version.